src/core.jl

```diff
@@ -0,0 +1,170 @@
+"""
+Type representing any object drawable on image
+"""
+abstract Drawable
+
+"""
+    p = Point(x,y)
+    p = Point(c)
+
+A `Drawable` point on the image
+"""
+immutable Point <: Drawable
+    x::Int
+    y::Int
+end
+
+abstract Line <: Drawable
+abstract Circle <: Drawable
+
+
+"""
+    line = LineTwoPoints(p1, p2)
+
+A `Drawable` infinite length line passing through the two points
+`p1` and `p2`.
+"""
+immutable LineTwoPoints <: Line
+    p1::Point
+    p2::Point
+end
+
+"""
+    line = LineNormal(ρ, θ)
+
+A `Drawable` infinte length line having perpendicular length `ρ` from
+origin and angle `θ` between the perpendicular and x-axis
+"""
+immutable LineNormal <: Line
+    ρ::Real
+    θ::Real
```

Comment:
Performance, yes. Consider this toy example

```julia
julia> immutable Point
           x::Real
       end

julia> f(p, i) = p.x * i

julia> p = Point(0.7)
Point(0.7)

julia> @code_llvm f(p, 0.7)

define %jl_value_t* @julia_f_72039(%jl_value_t*, double) #0 {
top:
  %ptls_i8 = call i8* asm "movq %fs:0, $0;\0Aaddq $$-2672, $0", "=r,~{dirflag},~{fpsr},~{flags}"() #2
  %ptls = bitcast i8* %ptls_i8 to %jl_value_t***
  %2 = alloca [5 x %jl_value_t*], align 8
  %.sub = getelementptr inbounds [5 x %jl_value_t*], [5 x %jl_value_t*]* %2, i64 0, i64 0
  %3 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %2, i64 0, i64 2
  %4 = bitcast %jl_value_t** %3 to i8*
  call void @llvm.memset.p0i8.i32(i8* %4, i8 0, i32 24, i32 8, i1 false)
  %5 = bitcast [5 x %jl_value_t*]* %2 to i64*
  store i64 6, i64* %5, align 8
  %6 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %2, i64 0, i64 1
  %7 = bitcast i8* %ptls_i8 to i64*
  %8 = load i64, i64* %7, align 8
  %9 = bitcast %jl_value_t** %6 to i64*
  store i64 %8, i64* %9, align 8
  store %jl_value_t** %.sub, %jl_value_t*** %ptls, align 8
  %10 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %2, i64 0, i64 4
  %11 = getelementptr [5 x %jl_value_t*], [5 x %jl_value_t*]* %2, i64 0, i64 3
  %12 = bitcast %jl_value_t* %0 to i64*
  %13 = load i64, i64* %12, align 8
  store %jl_value_t* inttoptr (i64 140216341607648 to %jl_value_t*), %jl_value_t** %3, align 8
  %14 = bitcast %jl_value_t** %11 to i64*
  store i64 %13, i64* %14, align 8
  %15 = call %jl_value_t* @jl_gc_pool_alloc(i8* %ptls_i8, i32 1432, i32 16)
  %16 = getelementptr inbounds %jl_value_t, %jl_value_t* %15, i64 -1, i32 0
  store %jl_value_t* inttoptr (i64 140216337464912 to %jl_value_t*), %jl_value_t** %16, align 8
  %17 = bitcast %jl_value_t* %15 to double*
  store double %1, double* %17, align 8
  store %jl_value_t* %15, %jl_value_t** %10, align 8
  %18 = call %jl_value_t* @jl_apply_generic(%jl_value_t** %3, i32 3)
  %19 = load i64, i64* %9, align 8
  store i64 %19, i64* %7, align 8
  ret %jl_value_t* %18
}
```

You will see the output of the LLVM code involved the garbage collector

```
  %15 = call %jl_value_t* @jl_gc_pool_alloc(i8* %ptls_i8, i32 1432, i32 16)
```

and use dynamic dispatch

```
  %18 = call %jl_value_t* @jl_apply_generic(%jl_value_t** %3, i32 3)
```

While the following is a lot simpler

```julia
julia> immutable Pointf
           x::Float64
       end

julia> p = Pointf(0.7)
Pointf(0.7)

julia> @code_llvm f(p, 0.7)

define double @julia_f_72066(%Pointf*, double) #0 {
top:
  %2 = getelementptr inbounds %Pointf, %Pointf* %0, i64 0, i32 0
  %3 = load double, double* %2, align 8
  %4 = fmul double %3, %1
  ret double %4
}
```